In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pickle

In [2]:
from ydata_profiling import ProfileReport

In [3]:
df = pd.read_csv("Advertising.csv")
df.head()

TV  Radio  Newspaper  Sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3   12.0
3  151.5   41.3       58.5   16.5
4  180.8   10.8       58.4   17.9

In [4]:
tools = ProfileReport(df)
tools.to_widgets()

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render widgets:   0%|          | 0/1 [00:00<?, ?it/s]

In [5]:
# Prepare the data
x = df.drop("Sales", axis=1) # Feature (X)
y = df["Sales"]              # Labesl  (Y)

In [6]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

In [7]:
model.fit(x, y)

LinearRegression()

eqn of line : `y = mx+c`

In [8]:
model.intercept_

4.62512407880865

In [9]:
model.coef_

array([0.05444578, 0.10700123, 0.00033566])

y = 0.05444578 TV + 0.10700123 Radio + 0.00033566 Newspaper <br>
`y = m1x1 + m2x2 + m3x3 + c`

There are three features, and the values 0.05444578, 0.10700123, and 0.00033566 represent the coefficients for each feature. These values indicate the relationship between each feature and the target variable in the model. Specifically, a positive coefficient means that an increase in that feature's value is associated with an increase in the target variable, while a negative coefficient means that an increase in that feature's value is associated with a decrease in the target variable.

As we can see that the Newspaper column is not participating(Showing much Linear Property), it's meant that x-axis(Investment) increasing stil there no direct impact in the y-axis(Sales). So we can just `ignore Newspaper` as it worth the investment.

but on the other hand, if we look at the data of Radio and TV it's giving us some return.

In [10]:
model.score(x, y)

0.9025912899684558

In [11]:
# Prepare the data
x2 = df.drop("Sales", axis=1) # Feature (X)
y2 = df["Sales"]              # Labesl  (Y)

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
model2 = LinearRegression()
model2.fit(x2, y2)
model2.score(x2, y2)

0.9025912899684558

Here we can see, even though we have not used the column Newspaper, there still not much effect on our score parameters.

We can call it `Feature Selection`.

In [14]:
import statsmodels.formula.api as smf
lm = smf.ols(formula='Sales ~ TV', data=df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.812
Model:                            OLS   Adj. R-squared:                  0.811
Method:                 Least Squares   F-statistic:                     856.2
Date:                Thu, 23 Feb 2023   Prob (F-statistic):           7.93e-74
Time:                        01:02:47   Log-Likelihood:                -448.99
No. Observations:                 200   AIC:                             902.0
Df Residuals:                     198   BIC:                             908.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.9748      0.323     21.624      0.000       6.339       7.611
TV             0.0555      0.002     29.260      0.000       0.052       0.059
==============================================================================
Omnibus:                        0.013   Durbin-Watson:                   2.029
Prob(Omnibus):                  0.993   Jarque-Bera (JB):                0.043
Skew:                          -0.018   Prob(JB):                        0.979
Kurtosis:                       2.938   Cond. No.                         338.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [15]:
import statsmodels.formula.api as smf
lm = smf.ols(formula='Sales ~ TV + Radio', data=df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.903
Model:                            OLS   Adj. R-squared:                  0.902
Method:                 Least Squares   F-statistic:                     912.7
Date:                Thu, 23 Feb 2023   Prob (F-statistic):          2.39e-100
Time:                        01:02:47   Log-Likelihood:                -383.34
No. Observations:                 200   AIC:                             772.7
Df Residuals:                     197   BIC:                             782.6
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.6309      0.290     15.952      0.000       4.058       5.203
TV             0.0544      0.001     39.726      0.000       0.052       0.057
Radio          0.1072      0.008     13.522      0.000       0.092       0.123
==============================================================================
Omnibus:                       16.227   Durbin-Watson:                   2.252
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.973
Skew:                          -0.434   Prob(JB):                     8.43e-07
Kurtosis:                       4.613   Cond. No.                         425.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [16]:
import statsmodels.formula.api as smf
lm = smf.ols(formula='Sales ~ TV + Radio + Newspaper', data=df).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.903
Model:                            OLS   Adj. R-squared:                  0.901
Method:                 Least Squares   F-statistic:                     605.4
Date:                Thu, 23 Feb 2023   Prob (F-statistic):           8.13e-99
Time:                        01:02:47   Log-Likelihood:                -383.34
No. Observations:                 200   AIC:                             774.7
Df Residuals:                     196   BIC:                             787.9
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      4.6251      0.308     15.041      0.000       4.019       5.232
TV             0.0544      0.001     39.592      0.000       0.052       0.057
Radio          0.1070      0.008     12.604      0.000       0.090       0.124
Newspaper      0.0003      0.006      0.058      0.954      -0.011       0.012
==============================================================================
Omnibus:                       16.081   Durbin-Watson:                   2.251
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               27.655
Skew:                          -0.431   Prob(JB):                     9.88e-07
Kurtosis:                       4.605   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

 Conclusion:
 
 here we can see that our P-value is Newspaper is 0.954, which it's predicted value is true only for 1-0.954, which is 0.046( in percent 4.6%), and other time false (95.4%)

In the first model, the R-squared is 0.903 and the adjusted R-squared is 0.902. This suggests that TV and Radio are both strong predictors of Sales.

In the second model, the R-squared is also 0.903 but the adjusted R-squared is 0.901. This suggests that adding the Newspaper predictor does not improve the model much, and may not be a strong predictor of Sales.